Fluid API v1

PepperCode1 opened the pull request:
This module adds a new FabricFlowableFluid class and FabricFluidBlock class. The new fluid class lets the user register one fluid instead of both a still fluid and a flowing fluid the vanilla way. It also lets the user set the maximum level of the fluid via an abstract method. Both classes provide some optimizations over vanilla's counterpart classes and contain some new methods that can be overriden for more control.

PepperCode1 committed:
Add Fluid Extensions Module

PepperCode1 committed:
Add Licenses

Add Licenses
Fix some formatting

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluid.java:
```diff
@@ -0,0 +1,110 @@
+	
+	/**
+	 * If a fluid is still, its state index is zero.
+	 * If a fluid is flowing, its state index is non-zero and represents its level and height.
```
Could you elaborate further on what part of the state index is level and which part is height?

PepperCode1 replied:
In the newer commits I removed the "and height" part from the javadoc, since that is confusing. The height of the fluid is level/(maxLevel+1), as seen in FlowableFluid#getHeight.

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluid.java:
```diff
@@ -0,0 +1,110 @@
+	@Override
+	protected void appendProperties(StateManager.Builder<Fluid, FluidState> builder) {
+		stateIndexProperty = IntProperty.of("state_index", 0, getLevels());
+        builder.add(stateIndexProperty);
```
this line needs to be indented with tabs

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluid.java:
```diff
@@ -0,0 +1,110 @@
+		if (falling) {
+			return getStill().getDefaultState().with(stateIndexProperty, ((FlowableFluidExtensions) this).getMaxLevel());
+		} else {
+			return getStill().getDefaultState();
```
you can drop the redundant else here and just return at the end of the method the default state.

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluid.java:
```diff
@@ -0,0 +1,110 @@
+import net.minecraft.state.property.IntProperty;
+
+public abstract class FabricFlowableFluid extends FlowableFluid {
+	private IntProperty stateIndexProperty;
```
This int property is never set?

Should also be final (if possible)

PepperCode1 replied:
It's not possible to make it final because it's set in the appendProperties method.

i509VCB replied:
OK I see.

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluidBlock.java:
```diff
@@ -0,0 +1,161 @@
+	public FabricFlowableFluidBlock(Settings settings) {
+		super(settings);
+		this.fluid = getFluid();
+		stateIndexProperty = this.fluid.getStateIndexProperty();
```
The getStateIndexProperty should be wrapped with an `Objects.requireNonNull`

PepperCode1 replied:
This value cannot be null unless mixins/asm are used. Should I still wrap it?

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluidBlock.java:
```diff
@@ -0,0 +1,161 @@
+			world.setBlockState(pos, Blocks.AIR.getDefaultState(), 11);
+			return fluid;
+		} else {
+			return Fluids.EMPTY;
```
Redundant else block

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluidBlock.java:
```diff
@@ -0,0 +1,161 @@
+
+	@Override
+	protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
+		builder.add(getFluid().getStateIndexProperty());
```
null check the state index property with an `Objects.requireNonNull`

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/api/fluid/extension/v1/FabricFlowableFluidBlock.java:
```diff
@@ -0,0 +1,161 @@
+	
+	public FabricFlowableFluidBlock(Settings settings) {
+		super(settings);
+		this.fluid = getFluid();
```
Null check the fluid

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+	
+	@ModifyConstant(
+		method = {
+			"getUpdatedState(Lnet/minecraft/world/WorldView;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;)Lnet/minecraft/fluid/FluidState;"
```
No need for the brackets here around the method descriptor

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+		method = {
+			"getUpdatedState(Lnet/minecraft/world/WorldView;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;)Lnet/minecraft/fluid/FluidState;"
+		},
+		constant = {@Constant(intValue=8)})
```
No need for brackets around the `@Constant`

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+			"getUpdatedState(Lnet/minecraft/world/WorldView;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;)Lnet/minecraft/fluid/FluidState;"
+		},
+		constant = {@Constant(intValue=8)})
+	public int onValue1(int value) {
```
Method name should be probably `modifyMaxFluidLevel`

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+			"getHeight(Lnet/minecraft/fluid/FluidState;)F",
+			"getShape(Lnet/minecraft/fluid/FluidState;Lnet/minecraft/world/BlockView;Lnet/minecraft/util/math/BlockPos;)Lnet/minecraft/util/shape/VoxelShape;"
+		},
+		constant = {@Constant(floatValue=9.0f)})
```
No need for Brackets around `@Constant`

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+			"getShape(Lnet/minecraft/fluid/FluidState;Lnet/minecraft/world/BlockView;Lnet/minecraft/util/math/BlockPos;)Lnet/minecraft/util/shape/VoxelShape;"
+		},
+		constant = {@Constant(floatValue=9.0f)})
+	public float onValue2(float value) {
```
Name this one as well

i509VCB commented on fabric-fluid-extensions-v1/src/main/java/net/fabricmc/fabric/mixin/fluid/extension/FlowableFluidMixin.java:
```diff
@@ -0,0 +1,95 @@
+		method = {
+			"getVelocity(Lnet/minecraft/world/BlockView;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/fluid/FluidState;)Lnet/minecraft/util/math/Vec3d;"
+		},
+		constant = {@Constant(floatValue=0.8888889f)})
```
I assume you get the point now. If it's one entry and the input object is an array, you can pass a single object in an annotation.

PepperCode1 committed:
Fix Everything

Rename to Fluid Extension API
Fix license and style
Add suggestions from review

sylv256 commented:
# ~~i have *[some serious]* concerns~~

1. ~~where is fluid volume api~~
1. ~~does it use millibuckets or millitoffets?~~

nevermind

Prospector commented:
Imo, this module should just be called `fabric-fluid-api-v1` so other future stuff like adding more tags for fluid behavior (such as for swimming, drowning, etc.) can be included in a more general fluid module, much like we have for items (`fabric-item-api-v1`).

A fluid volume api would be `fabric-fluid-volume-api-v1`

PepperCode1 committed:
Rename Module and Fix Test Mod

Renamed the entire module and any instance of "fabric-fluid-extension-api" to "fabric-fluid-api"
Removed a comma in the fabric.mod.json of the test mod

PepperCode1 commented:
I have renamed the module as per Prospector's suggestion.

Virtuoel commented on fabric-fluid-api-v1/src/main/java/net/fabricmc/fabric/api/fluid/v1/FabricFlowableFluid.java:
```diff
@@ -0,0 +1,108 @@
+
+	@Override
+	protected void appendProperties(StateManager.Builder<Fluid, FluidState> builder) {
+		stateIndexProperty = IntProperty.of("state_index", 0, getLevels());
```
This current approach would result in duplicated property instances when there are multiple instances of `FabricFlowableFluid` that have the same amount of levels as each other. The index property is also not accessible without having an instance of `FabricFlowableFluid` or `FabricFlowableFluidBlock` to call `getStateIndexProperty` on.  

Properties should be relatively easily accessible from a static context and also have no duplicated instances, with the same property instance being shared among all blocks or fluids that use it, similar to how the vanilla ones in `Properties` are all static with e.g. the 6-way direction property being one instance shared between all 6-way directional blocks.  


I suppose one approach could be a method that takes a level and returns the corresponding property from a map of levels to properties, or an array where the index is the levels, or some other data structure, with it first creating the property for the level if it doesn't exist yet.  

(Don't know the best approach or naming for something like that which would also fit with the rest of Fabric API. Maybe some sort of `FabricProperties.INSTANCE.getFluidIndexFromLevels` method for lazily creating/getting the corresponding property? Not sure.)  

Anyways, with properties, they should not have duplicate instances and should be easily accessible in a static context.

PepperCode1 committed:
Merge branch '1.16' into 1.16-fluid-extensions-api

PepperCode1 committed:
Clean up a few things

PepperCode1 commented:
I am unsure about this PR because my original goal with it was to essentially replace vanilla's fluid system (which IMO isn't very good), and this PR accomplishes this quite well. I am now realizing though, that forcing users to use a single fluid and the custom fluid block class takes away control, and that this PR might not wholly fit into Fabric API. In the Discord API channel, it was discussed that a unified fluid class does not fit into Fabric API and that the FluidBlock class should be mixed into instead of replaced, but I am really not sure how to improve the system only partially.
